### ISSUE TITLE
Better syntax for fixed effects?

### ISSUE THREAD

**Issue by pdeffebach** [JuliaStats/StatsModels.jl#196]
Currently if you have a variable whose levels you want to turn into dummy variables you need to put them as a categorical variable or string. This creates ambiguity

```
@formula(income ~ educ + state)
```

what if `state` is coded as `1, 2, ..., 50` rather than `"AL", "AK", ..., "WY"`?  This will cause an error. Additionally, it's ambiguous at the call-site. The reader doesn't know which variables are categorical and which aren't. 

Can we have a syntax like 

```
@formula(income ~ educ + fe(state))
```

or `factor(state)` or similar to indicate when a variable should be treated as categorical?

**Comment by palday** [comment]
This is what the `contrasts` argument is for.  For example, you could use `contrasts=Dict(:state => DummyCoding())` to achieve that effect in `GLM.jl` or `MixedModels.jl`.

It's also generally very Julian to make sure that your data are typed appropriately; in this case that means making sure that the Table columns are the correct type. You can't know the meaning of anything without knowing the types involved.

Even when you know the programmatic types, it's still not possible to interpret a statistical model without knowing the units, how the data were collected, etc. Or in the case of `state` -- are the levels states in the United States? Couldn't they be states in Australia? You can't know at the call site. 

More generally, I would say that no single-line in a program can be understood on its own; it always exists carries the context of (at the very least) its local scope and potentially the global environment.

**Comment by kleinschmidt** [comment]
I do agree this would be useful to have.  Finding the right syntax is an issue.  The other precedent I'm aware of for this is R, where `C(term, [contrasts_type])` can be used inside a formula to create a categorical variable.

In the Terms 2.0 era, this information is most naturally represented in the schema, and the API is that a term-to-contrasts (or just `CategoricalTerm`) is passed as the optional `hints` argument to the `schema` functions.  That's where the `contrasts=` kwarg in GLM/MixedModels ends up.

**Comment by kleinschmidt** [comment]
We could, for instance, have functions like `dummy`, `effects`, `herlmert`, `seqdiff` which can be used in a formula to create categorical terms with that coding.  Or allow constructors for contrasts to be used in the formula, which is elegant but punny because there's actually no way to know whether the call is to a constructor of a type that's `<:AbstractContrasts`.

The other alternative is to have some single function (like `fe`, or `C` from R, or `categorical`) which takes a term name and an optional contrasts type/constructor call.

**Comment by pdeffebach** [comment]
In `fixest` you can do `factor` to specify to treat a column as categorical. I always use it where possible, even if there is a slight performance hit. 

I would really rather this be in the `@formula` call because it's part of a transformation that creates the model matrix, and thus ought to be handled by StatsModels rather than a keyword argument, which would indicate it's being handled by GLM. 



> The other alternative is to have some single function (like `fe`, or `C` from R, or `categorical`) which takes a term name and an optional contrasts type/constructor call.

Could you clarify this? I'm fine with somewhat more verbose syntax, as long as it's in the `@formula` call. 

Another parallel here is Stata, which *requires* you to specify if a variable should be treated as a fixed effects. 

```
reg income educ i.state
```

**Comment by palday** [comment]
I like the idea of being able to set contrasts within a formula, but that's different from just vaguely marking something as categorical, which doesn't actually achieve the ostensible goal of making formulae interpretable without context. (I have a whole research project on misunderstanding contrasts and incorrect use of default contrasts, so if we're going to make these things explicit, we should do it in a way that encourages good behavior.)

That said, I would strongly avoid the "fixed effects" terminology. That has different meanings in different areas of applied statistics. For example, all of the examples given here whether interpreted as continuous or categorical are fixed effects in the vocabulary of mixed-effects models or even regression models (including meta-analysis) in general outside of econometrics.

I would also recommend having the data represented in a way that actually reflects its structure. If you manually code `state`s as numbers, that's helpful neither to the human (does not help readibility) nor the computer (regression contrasts don't correspond to the manual numeric coding; pooled/categorical array representations do their magic equally well regardless of whether the original datatype was string or integer). This really helps avoid such accidents as having a variable interpreted as numeric when it's not. Especially if you something obvious/interpretable at the call site, then `unique(state)` is only useful for interpreting the data if `state` is represented in an appropriate semantic way and not as an opaque mapping to integers.

**Comment by pdeffebach** [comment]
> That said, I would strongly avoid the "fixed effects" terminology. That has different meanings in different areas of applied statistics.

Sure, we don't need to call it "fixed effects". But what I mean, specifically, is treating a numeric variable as a categorical one at the call-site. 

> If you manually code `state`s as numbers, that's helpful neither to the human (does not help readibility) nor the computer

As an RA working on larger projects, I don't always have the chance to control the way the data is constructed all the time. I am proposing a convenient syntax for someone who routinely encounters this problem. 

I will also note that frequently there *is* ambiguity about when a variable should be treated as categorical or continuous. In event studies, for instance, you frequently want to use `year` as a continuous variable and then separately examine effects where years are treated as a categorical variable. 

This is a not-uncommon feature request, as evidenced by discourse thread [here](https://discourse.julialang.org/t/regression-fixed-effects-quantiles/45427/17?u=pdeffebach). Again, other software systems support this feature.

**Comment by palday** [comment]
This is diverging into one of my rants, but
> As an RA working on larger projects, I don't always have the chance to control the way the data is constructed all the time. I am proposing a convenient syntax for someone who routinely encounters this problem.

My advice is to immediately put the data into a semantic form upon loading, just like text bytes should immediately be decoded into their semantic representation (usually as Unicode code points) instead of a particular encoding. 

I think making contrast coding explicit is great. But, again, that's a step further than just saying "categorical" or "continuous". Because even categorical is generally not interpretable without knowing the contrasts, except at a level of abstraction where the continuous/categorical distinction is less relevant (e.g. at the level where it's generally "the influence of year" and not "how do I interpret the relevant model coefficients associated with `year`.) 

**tl;dr:** 

:+1:  for explicit contrast coding in formulae.  This could be useful in mixed-effects models. where the fixed and random effects could have different contrast schemes. 
:-1:  for `categorical` in formulae without contrast coding.

**Comment by pdeffebach** [comment]
I feel like there is still an inconsistency, though. If using `@formula(income ~ edu + state)` is satisfactory, to you, where state is a categorical variable, without extra detail about the type of encoding used, then why should `@formula(income ~ edu + categorical(state))`?

**Comment by palday** [comment]
All or nothing. Either you're willing to accept that a formula is interpreted in the context of its data and the code above it or you want the formula to be an explicit declaration of everything about the data's structure.

**Comment by Tokazama** [comment]
I think it would be good to have a this in the formula instead of having to use contrasts using something like `categorical(state)`. If `state` is numeric then it could to be used to indicate something is ordinal.

**Comment by palday** [comment]
>  If the reason we shouldn't add this is because the contrast information isn't explicitly available then we should probably not assume symbols or strings are categorical unless the data is a CategoricalArray or the contrast is explicitly provided.

This is actually how grouping/blocking variables are handled in MixedModels. These must be explicitly categorical whether through `CategoricalArray`, `PooledArray` or through an explicit contrast.

> But I don't want to store my Likert scales as strings or symbols. Sure, I can use CategoricalArrays to provide explicit levels, but not every modeling framework knows how to use these and I end up with a categorical and numeric instance saved. None of this is unmanageable, but it's needlessly messy and obnoxious.

This sounds like you treat them sometimes as numbers, sometimes as categorical. Exactly in that case, I think it matters even more to know what valid contrasts would be for the categorical case. Again, I'm going back to the core claim of "obviousness at the call site". Telling me that your variable is categorical doesn't tell me enough to understand what the resulting model matrix looks like -- I need to know the contrasts for that. If you don't need to know exactly what the model matrix looks like, then I'm not sure that adding `categorical` to the formula gives you any more information than the 'bare' variant which tells you which main effects and interactions are present. `categorical` without contrasts (or some indicator of the (number of) levels) doesn't even tell you how many columns you have.

For Likert in particular, you  could define a `Likert{nlevels}` type with appropriate comparison operators that stores the ratings internally as an integer. This is then a semantic form that captures that the Likert values are in some sense integers, but that not all ratio-scale or perhaps even interval-scale operations (in the sense of [Stevens 1946](doi.org/10.1126/science.103.2684.677)) are valid.

**Comment by Tokazama** [comment]
> This is actually how grouping/blocking variables are handled in MixedModels. These must be explicitly categorical whether through CategoricalArray, PooledArray or through an explicit contrast.

Variables that are `String` are considered categorical in MixedModels (at least in the version I use). The formula just assumes dummy coding unless some other contrast is provided. Why couldn't you just do the same with other variables that are declared as "categorical", assume dummy coding unless told otherwise?

> For Likert in particular, you could define a Likert{nlevels} type

This assumes I'm still treating them as a Likert scale for something like an ordered logit model. There are other things that I do with them that are much more complicated and redefining all the `Number` methods so I can do this seems like overkill. The specifics of what I'm doing aren't really important for the current discussion. It was simply an example to illustrate how the current tools available aren't as flexible to different workflows as they could be.

**Comment by palday** [comment]
> Why couldn't you just do the same with other variables that are declared as "categorical", assume dummy coding unless told otherwise?

This is really what I'm getting at. If `y ~ categorical(x)` implies `x` is dummy coded, then why not just have `y ~ dummycoding(x)`? It's explicit and clear.  If you need different contrasts with `y ~ categorical(x)` ... `contrasts=Dict(:x => MyCoding())`, then `y ~ mycoding(x)` is also convenient.  In other words, what does `categorical` in the formula syntax get you that `dummycoding` doesn't if `categorical` maps to `dummycoding` anyway when contrasts aren't specified (which would make the `categorical` marker moot)?
